workflow/controller/controller.go

```diff
@@ -529,19 +529,20 @@ func (wfc *WorkflowController) processNextPodCleanupItem(ctx context.Context) bo
 					Value:     "true",
 				},
 			}
-			obj, _, _ := wfc.podInformer.GetIndexer().GetByKey(namespace + "/" + podName)
-			if p, ok := obj.(*apiv1.Pod); ok {
-				i := slices.Index(p.Finalizers, common.Finalizer)
-				if i >= 0 {
-					patch = append(patch, common.DeleteFinalizerPatch{
-						Operation: "remove",
-						Path:      fmt.Sprintf("/metadata/finalizers/%d", i),
-					})
-				}
+			p, err := wfc.getPodFromApi(ctx, namespace, podName)
```

Comment:
I checked the difference between the SharedInformer and the Kubernetes API pod information when I reproduced the flake test locally.

<details>
<summary> code for check </summary>

```diff
diff --git a/workflow/controller/controller.go b/workflow/controller/controller.go
index 364f51aef..66e90fcc3 100644
--- a/workflow/controller/controller.go
+++ b/workflow/controller/controller.go
@@ -530,7 +530,18 @@ func (wfc *WorkflowController) processNextPodCleanupItem(ctx context.Context) bo
                                },
                        }
                        obj, _, _ := wfc.podInformer.GetIndexer().GetByKey(namespace + "/" + podName)
+                       pod, err := wfc.kubeclientset.CoreV1().Pods(namespace).Get(ctx, podName, metav1.GetOptions{})
+                       if err != nil {
+                               log.WithError(err).Warn("failed to get latest pod by name")
+                               return err
+                       }
+                       log.Info("!!!!!!!!!!!!!!!!!!!!!!!!!")
+                       log.Info("Finalizers from api")
+                       log.Info(pod.Finalizers)
                        if p, ok := obj.(*apiv1.Pod); ok {
+                               log.Info("!!!!!!!!!!!!!!!!!!!!!!!!!")
+                               log.Info("Finalizers from cache")
+                               log.Info(p.Finalizers)
                                i := slices.Index(p.Finalizers, common.Finalizer)
                                if i >= 0 {
                                        patch = append(patch, common.DeleteFinalizerPatch{
@@ -541,7 +552,7 @@ func (wfc *WorkflowController) processNextPodCleanupItem(ctx context.Context) bo
                        }
                        data, _ := json.Marshal(patch)
                        log.WithField("data", string(data)).Debug("patching pod")
-                       _, err := pods.Patch(
+                       _, err = pods.Patch(
                                ctx,
                                podName,
                                types.JSONPatchType,
```

</details>

Here is an excerpt from the log. You can see that the SharedInformer pod information has the finalizer, but the Kubernetes API does not.

```
controller: time="2023-11-20T08:14:40.783Z" level=info msg="cleaning up pod" action=labelPodCompleted key=argo/test-pod-cleanup-on-workflow-completion-label-not-match-h6mfm/labelPodCompleted
controller: time="2023-11-20T08:14:40.783Z" level=info msg="insignificant pod change" key=argo/test-pod-cleanup-on-workflow-completion-label-not-match-h6mfm
controller: time="2023-11-20T08:14:40.785Z" level=info msg="!!!!!!!!!!!!!!!!!!!!!!!!!"
controller: time="2023-11-20T08:14:40.785Z" level=info msg="Finalizers from api"
controller: time="2023-11-20T08:14:40.785Z" level=info msg="[]"
controller: time="2023-11-20T08:14:40.785Z" level=info msg="!!!!!!!!!!!!!!!!!!!!!!!!!"
controller: time="2023-11-20T08:14:40.785Z" level=info msg="Finalizers from cache"
controller: time="2023-11-20T08:14:40.785Z" level=info msg="[workflows.argoproj.io/status]"
controller: time="2023-11-20T08:14:40.788Z" level=warning msg="failed to clean-up pod" action=labelPodCompleted error="the server rejected our request due to an error in our request" key=argo/test-pod-cleanup-on-workflow-completion-label-not-match-h6mfm/labelPodCompleted
controller: time="2023-11-20T08:14:40.788Z" level=warning msg="Non-transient error: the server rejected our request due to an error in our request"
controller: time="2023-11-20T08:14:41.698Z" level=info msg="cleaning up pod" action=killContainers key=argo/test-pod-cleanup-on-workflow-completion-label-not-match-h6mfm/killContainers
```

The `the server rejected our request due to an error in our request"` can be reproduced if the pod doesn't have the finalizer and try applying the remove finalizer patch.

<details>
<summary> reproduce the server rejected error </summary>

```shell
$ vi hello-world-no-finalizer.yaml
------------
apiVersion: v1
kind: Pod
metadata:
  generateName: hello-world-
spec:
  containers:
  - image: hello-world
    name: hello-world
    imagePullPolicy: IfNotPresent
  restartPolicy: Never
------------
$ kubectl create -f hello-world-no-finalizer.yaml
$ kubectl patch pod hello-world-z5xt7 --type=json -p '[{"op":"remove","path":"/metadata/finalizers/0"}]'
The request is invalid: the server rejected our request due to an error in our request
```

</details>

If the pod has the finalizer, the patch will be successful.

<details>
<summary> success patch </summary>

```shell
$ vi hello-world-finalizer.yaml
------------
apiVersion: v1
kind: Pod
metadata:
  generateName: hello-world-
  finalizers:
  - workflows.argoproj.io/status
spec:
  containers:
  - image: hello-world
    name: hello-world
    imagePullPolicy: IfNotPresent
  restartPolicy: Never
------------
$ kubectl create -f hello-world-finalizer.yaml
$ kubectl patch pod hello-world-cl8f7 --type=json -p '[{"op":"remove","path":"/metadata/finalizers/0"}]'
pod/hello-world-cl8f7 patched
```

</details>

Finally, when the server rejected error happens,  the pod moves to the killContainers action, and `parsePodCleanupKey` tries to `wfc.signalContainers`, but it got stuck and timed out(I don't know why, but I think this is a different scope).